libraries/AP_Airspeed/AP_Airspeed_MS5525.cpp

```diff
@@ -0,0 +1,286 @@
+{
+    const uint8_t addresses[] = { MS5525D0_I2C_ADDR_1, MS5525D0_I2C_ADDR_2 };
+    bool found = false;
+    for (uint8_t i=0; i<ARRAY_SIZE(addresses); i++) {
+        dev = hal.i2c_mgr->get_device(get_bus(), addresses[i]);
+        if (!dev) {
+            continue;
+        }
+        if (!dev->get_semaphore()->take(0)) {
+            continue;
+        }
+
+        // lots of retries during probe
+        dev->set_retries(5);
+
+        found = read_prom();
+        
+        if (found) {
+            printf("MS5525: Found sensor on bus %u address 0x%02x\n", get_bus(), addresses[i]);
+            break;
+        }
+        dev->get_semaphore()->give();
+    }
+    if (!found) {
+        printf("MS5525: no sensor found\n");
+        return false;
+    }
+
+    // Send a command to read temperature first
+    uint8_t reg = REG_CONVERT_TEMPERATURE;
+    dev->transfer(&reg, 1, nullptr, 0);
+    state = 0;
+
+    // drop to 2 retries for runtime
+    dev->set_retries(2);
+
+    dev->get_semaphore()->give();
+
+    // read at 80Hz
+    dev->register_periodic_callback(1000000UL/80U,
+                                    FUNCTOR_BIND_MEMBER(&AP_Airspeed_MS5525::timer, bool));
+    return true;
+}
+
+
+/**
+ * CRC used by MS pressure devices
+ */
+uint16_t AP_Airspeed_MS5525::crc4_prom(void)
+{
+    uint16_t n_rem = 0;
+    uint8_t n_bit;
+
+    for (uint8_t cnt = 0; cnt < sizeof(prom); cnt++) {
+        /* uneven bytes */
+        if (cnt & 1) {
+            n_rem ^= (uint8_t)((prom[cnt >> 1]) & 0x00FF);
+        } else {
+            n_rem ^= (uint8_t)(prom[cnt >> 1] >> 8);
+        }
+
+        for (n_bit = 8; n_bit > 0; n_bit--) {
+            if (n_rem & 0x8000) {
+                n_rem = (n_rem << 1) ^ 0x3000;
+            } else {
+                n_rem = (n_rem << 1);
+            }
+        }
+    }
+
+    return (n_rem >> 12) & 0xF;
+}
+
+bool AP_Airspeed_MS5525::read_prom(void)
+{
+    // reset the chip to ensure it has correct prom values loaded
+    uint8_t reg = REG_RESET;
+    if (!dev->transfer(&reg, 1, nullptr, 0)) {
+        return false;
+    }
+    hal.scheduler->delay(5);
+
+    bool all_zero = true;
+    for (uint8_t i = 0; i < 8; i++) {
+        if (!dev->read_uint16_be(REG_PROM_BASE+i*2, prom[i])) {
+            return false;
+        }
+        if (prom[i] != 0) {
+            all_zero = false;
+        }
+    }
+
+    if (all_zero) {
+        return false;
+    }
+
+    /* save the read crc */
+    const uint16_t crc_read = prom[7] & 0xf;
+
+    /* remove CRC byte */
+    prom[7] &= 0xff00;
+
+    uint16_t crc_calc = crc4_prom();
+    if (crc_read != crc_calc) {
+        printf("MS5525: CRC mismatch 0x%04x 0x%04x\n", crc_read, crc_calc);
+    }
+    return crc_read == crc_calc;
+}
+
+
+/*
+  read from the ADC
+ */
+int32_t AP_Airspeed_MS5525::read_adc()
+{
+    uint8_t val[3];
+    if (!dev->read_registers(REG_ADC_READ, val, 3)) {
+        return 0;
+    }
+    return (val[0] << 16) | (val[1] << 8) | val[2];
```

Comment:
I don't like the cross-your-fingers promotion relying on promotion order. I think it's easier to read if we cast just for our own sanity.